### PR TITLE
[W08H02] test(TaskFunctionBehaviorTest): Add strict equals comparison

### PR DESCRIPTION
**Pull request by hmelder**
> Überschriebene Methoden hashCode und equals, die jeweils nur von ID abhängig sein sollen.

The goal of the first test is to see if TaskFunction::equals uses a Class::getClass comparison thus locking Class::equals to instances of the same class. TaskFunction::equals may only be dependent on ID, in case of altering a different instance of TaskFunction to the same value (e.g. by a method overwrite) it should yield true.

The second test is trivial as we are not overwriting any methods in the subclass, we could still produce an undefined behaviour in flawed implementations that can't handle subclasses.

### CODE REVIEW

**Comment by hmelder** [comment]
Thinking about it, you should not be able to change the internal ID as equals compares the field (in this case ID). When calling equals, we should always compare the IDs of TaskFunction even if there is a subclass overwriting ID and its accessor (Variable hiding).

TLDR: You should not access ID via getID for comparison as it may be altered. This in turn changes the symmetry of the equality method (y.equals(x) and x.equals(y) where y is a subclass with getID overwritten and equals calls getID instead of this.ID).

So while the test case I wrote cannot be reproduced in real-world conditions (without using reflection) we should check whether equals calls getID.

**Comment by hmelder** [comment]
This should be it.

File: w08h02/test/pgdp/pools/TaskFunctionBehaviorTest.java [hunk 1]
```diff
@@ -143,4 +143,25 @@ void equalsTwoDifferentTaskFunctions() {
...
+
+            @Override
+            // In case TaskFunction::equals uses hashCode() which changes symmetry of equals
+            public int hashCode() {
+                return func1.hashCode();
+            }
+        };
+
+        assertFalse(func1.equals(func2), "equals should yield false for task with overwritten methods");
+    }
```
Comment (nilsreichardt) [inline w08h02/test/pgdp/pools/TaskFunctionBehaviorTest.java:165]: Are you sure that we should have getter for ID?

Comment (hmelder) [reply w08h02/test/pgdp/pools/TaskFunctionBehaviorTest.java:165]: They have not explicitly mentioned it, but it would make sense. I can leave the getID() there but extract the value by reflection. So we can still test for potential problems for people that have implemented the getter.

**Review by MaximilianAnzinger — APPROVED** [review]

File: w08h02/test/pgdp/pools/TaskFunctionBehaviorTest.java (hunk 1 above)
Comment (ipadmini22) [reply w08h02/test/pgdp/pools/TaskFunctionBehaviorTest.java]: That @Override for the getID must be removed anyway, in case that function does not exist. Getter/Setter are not required and cannot be expected. They are also not expected by public tests.

Comment (hmelder) [reply w08h02/test/pgdp/pools/TaskFunctionBehaviorTest.java]: > That @OverRide for the getID must be removed anyway, in case that function does not exist. Getter/Setter are not required and cannot be expected. They are also not expected by public tests. Generally, I think those tests are really useful!

Thanks for pointing that out.

**Review by ipadmini22 — APPROVED** [review]
I guess we can merge! LGTM

File: w08h02/test/pgdp/pools/TaskFunctionBehaviorTest.java [hunk 2]
```diff
@@ -143,4 +143,35 @@ void equalsTwoDifferentTaskFunctions() {
         assertFalse(task1.equals(task2), "equals should yield false for two different TaskFunction instances");
     }
 
+    @Test
+    @DisplayName("equals() should return false for identical input and manipulated subclassed function")
+    void equalsUnderlyingSubclassedFunction() {
+        final int idOfFunc1;
+        var func1 = new TaskFunction<Integer, Integer>(SQUARE);
```
Comment (nilsreichardt) [inline w08h02/test/pgdp/pools/TaskFunctionBehaviorTest.java:150]:
```suggestion
        var func1 = new TaskFunction<Integer, Integer>(FunctionLib.SQUARE);
```

Is this missing?

Comment (ipadmini22) [reply w08h02/test/pgdp/pools/TaskFunctionBehaviorTest.java:150]: Works, because those imports exist
```import static pgdp.pools.FunctionLib.SQUARE;
import static pgdp.pools.FunctionLib.SUM_OF_HALFS;```

**Comment by felixhauptmann** [comment]
LGTM

**Review by felixhauptmann — APPROVED** [review]
LGTM

File: w08h02/test/pgdp/pools/TaskFunctionBehaviorTest.java (hunk 2 above)
Comment (felixhauptmann) [reply w08h02/test/pgdp/pools/TaskFunctionBehaviorTest.java:150]: i just compiled it without any issues

**Comment by hmelder** [comment]
Would be good to have this in main before the deadline :)

File: w08h02/test/pgdp/pools/TaskFunctionBehaviorTest.java (hunk 2 above)
Comment (HendrikHuebner) [reply w08h02/test/pgdp/pools/TaskFunctionBehaviorTest.java:150]: I had no issues either

**Comment by HendrikHuebner** [comment]
@MaximilianAnzinger please merge this
